mqtt/mqtt3/src/client/mod.rs

```diff
@@ -810,6 +818,16 @@ impl Error {
             _ => false,
         }
     }
+
+    fn is_connection_error(&self) -> bool {
+        match self {
+            Error::DecodePacket(crate::proto::DecodeError::Io(_))
+            | Error::EncodePacket(crate::proto::EncodeError::Io(_))
+            | Error::ServerClosedConnection
+            | Error::PingTimer(_) => true,
+            _ => false,
```

Comment:
I don't think we have tests for all these cases now

Reply:
DecodePacket is not covered in tests, didn't find a way to test it. PingTimer was removed as it seems it is not used.